plugins/modules/source_control/gitlab/gitlab_project_variable.py

```diff
@@ -162,106 +241,147 @@ def list_all_project_variables(self):
         if self._module.check_mode:
-            return
+            return True
         var = {
-            "key": key, "value": value,
-            "masked": masked, "protected": protected,
-            "variable_type": variable_type,
+            "key": var_obj.get('key'), "value": var_obj.get('value'),
+            "masked": var_obj.get('masked'), "protected": var_obj.get('protected'),
+            "variable_type": var_obj.get('variable_type'), "environment_scope": var_obj.get('environment_scope')
         }
-        if environment_scope is not None:
-            var["environment_scope"] = environment_scope
-        return self.project.variables.create(var)
-
-    def update_variable(self, key, var, value, masked, protected, variable_type, environment_scope):
-        if (var.value == value and var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            return False
+        self.project.variables.create(var)
+        return True
 
+    def update_variable(self, var_obj):
         if self._module.check_mode:
             return True
+        self.delete_variable(var_obj)
+        self.create_variable(var_obj)
+        return True
 
-        if (var.protected == protected and var.masked == masked
-                and var.variable_type == variable_type
-                and (var.environment_scope == environment_scope or environment_scope is None)):
-            var.value = value
-            var.save()
+    def delete_variable(self, var_obj):
+        if self._module.check_mode:
             return True
-
-        self.delete_variable(key)
-        self.create_variable(key, value, masked, protected, variable_type, environment_scope)
+        self.project.variables.delete(var_obj.get('key'), filter={'environment_scope': var_obj.get('environment_scope')})
         return True
 
-    def delete_variable(self, key):
-        if self._module.check_mode:
-            return
-        return self.project.variables.delete(key)
+
+def compare(requested_variables, existing_variables, state):
+    # we need to do this, because it was determined in a previous version - more or less buggy
+    # basically it is not necessary and might results in more/other bugs!
+    # but it is required  and only relevant for check mode!!
+    # logic represents state 'present' when not purge. all other can be derived from that
+    # untouched => equal in both
+    # updated => name and scope are equal
+    # added => name and scope does not exist
+    untouched = list()
+    updated = list()
+    added = list()
+
+    if state == 'present':
+        existing_key_scope_vars = list()
+        for item in existing_variables:
+            existing_key_scope_vars.append({'key': item.get('key'), 'environment_scope': item.get('environment_scope')})
+
+        for idx in range(len(requested_variables)):
```

Comment:
```suggestion
        for var in requested_variables:
```
and replace `requested_variables[idx]` by `var` below.